File: lib/services/mileage_fee_decorator.dart
import '../models/rental.dart';
import 'price_calculator.dart';
import 'price_decorator.dart';

class MileageFeeDecorator extends PriceDecorator {
  final int mileage;
  final int pricePerKm;

  MileageFeeDecorator(
    PriceCalculator calculator, {
    required this.mileage,
    this.pricePerKm = 100,
  }) : super(calculator);

  @override
  int calculate(Rental rental) {
    int basePrice = calculator.calculate(rental);
    int mileageFee = mileage * pricePerKm;
    return basePrice + mileageFee;
  }
}

File: src/main/java/org/example/Payment/MileageFeeDecorator.java
package org.example.Payment;

public class MileageFeeDecorator extends PriceDecorator {
    private double mileage;
    private double ratePerKm;
    public MileageFeeDecorator(PriceCalculator priceCalculator, double mileage, double ratePerKm) {
        super(priceCalculator);
        this.mileage = mileage;
        this.ratePerKm = ratePerKm;
    }
    public int calculatePrice() {
        int basePrice = super.calculatePrice();
        int mileageFee = (int) (mileage*ratePerKm);
        System.out.println(" [결제 가격] 기본 대여료: " + basePrice + "원, 주행 거리(" + mileage + "): " + mileageFee);
        return basePrice + mileageFee;
    }
}
